### PR TITLE
#259 Support for node labels to set eniConfig

### PR DESCRIPTION
**Pull request by etopeter**
#259

Added support for node labels and unit tests.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by md2k** [comment]
Tested this PR with my EKS and it is awesome. probably for a while we going to use custom build for aws-node to ensure that we able to do automated cni configuration.

File: pkg/eniconfig/eniconfig.go [hunk 1]
```diff
@@ -115,10 +116,22 @@ func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
...
-				h.controller.eniLock.Lock()
-				defer h.controller.eniLock.Unlock()
-				h.controller.myENI = eniConfigDefault
-				log.Infof(" Setting myENI to: %s", eniConfigDefault)
+
+				label := o.GetLabels()
+
+				val, ok := label[eniConfigLabelDef]
+				if ok {
+					h.controller.eniLock.Lock()
```
Comment (nak3) [inline pkg/eniconfig/eniconfig.go]: There are three duplicated `Lock`, `Unlock` lines around this block. Would you like to refactor them?

For example, you can write under L106 as:

```
        case *corev1.Node:

                log.Infof("Handle corev1.Node: %s, %v, %v", o.GetName(), o.GetAnnotations(), o.GetLabels())
                if h.controller.myNodeName == o.GetName() {
                        val, ok := o.GetAnnotations()[eniConfigAnnotationDef]
                        if !ok {
                                val, ok = o.GetLabels()[eniConfigLabelDef]
                                if !ok {
                                        val = eniConfigDefault
                                }
                        }

                        h.controller.eniLock.Lock()
                        defer h.controller.eniLock.Unlock()
                        h.controller.myENI = val
                        log.Infof(" Setting myENI to: %s", val)
                }
        }
        return nil
```

File: pkg/eniconfig/eniconfig_test.go [hunk 2]
```diff
@@ -68,6 +68,33 @@ func updateNodeAnnotation(hdlr sdk.Handler, nodeName string, configName string,
 	hdlr.Handle(nil, event)
 }
 
+func updateNodeLabel(hdlr sdk.Handler, nodeName string, configName string, toDelete bool) {
```
Comment (nak3) [inline pkg/eniconfig/eniconfig_test.go]: This function is almost same with `updateNodeAnnotation()`. It is better to combine them?

File: pkg/eniconfig/eniconfig_test.go [hunk 3]
```diff
@@ -144,3 +171,44 @@ func TestNodeENIConfig(t *testing.T) {
 	assert.Equal(t, defaultCfg, *outputCfg)
 
 }
+
+func TestNodeENIConfigLabel(t *testing.T) {
```
Comment (nak3) [inline pkg/eniconfig/eniconfig_test.go]: Same with above comment. Is it not possible to merge this test into `TestNodeENIConfig()`?

**Comment by nak3** [comment]
I am not an official reviewer so please disregard my comments if you disagree.

File: pkg/eniconfig/eniconfig_test.go (hunk 2 above)
Comment (mattlandis) [reply pkg/eniconfig/eniconfig_test.go]: Can we extract the strictly duplicate code.  I think two methods is fine, if we remove the code that is the same toa  `createEvent` function.

File: pkg/eniconfig/eniconfig_test.go (hunk 3 above)
Comment (mattlandis) [reply pkg/eniconfig/eniconfig_test.go]: I'm planning a bigger refactor of the test code so I am OK with this.

**Comment by md2k** [comment]
when this can be merged?

**Comment by etopeter** [comment]
There is another PR with same functionality, more features and cleaner logic here [280](https://github.com/aws/amazon-vpc-cni-k8s/pull/280) So hope that one gets merged and you will also be able to set custom annotations and labels.

**Comment by mogren** [comment]
Can we cancel this PR since #280 has been merged?

**Pull request closed without merging**